src/blocks/timewarrior.rs

```diff
@@ -0,0 +1,193 @@
+//! Time and information of the current timewarrior task
+//!
+//! Clicking left mouse stops or resumes the task
+//!
+//! # Configuration
+//!
+//! Key | Values | Default
+//! ----|--------|--------
+//! `interval` | Update interval in seconds | `30`
+//! `format` | A string to customise the output of the block. See placeholders. | <code>" $icon {$elapsed &vert;}"</code>
+//! `info` | The threshold of minutes the task turns into a info state | -
+//! `good` | The threshold of minutes the task turns into a good state | -
+//! `warning` | The threshold of minutes the task turns into a warning state | -
+//! `critical` | The threshold of minutes the task turns into a critical state | -
+//!
+//! Placeholder | Value | Type | Unit
+//! ------------|-------|------|------
+//! `icon`   | A static icon | Icon | -
+//! `elapsed`| Elapsed time in format H:MM (Only present if task is active) | Text | -
+//! `tags`   | Tags of the active task separated by space (Only present if task is active) | Text | -
+//! `annotation` | Annotation of the active task (Only present if task is active) | Text | -
+//!
+//! Action          | Default button
+//! ----------------|----------------
+//! `stop_continue` | Left
+//!
+//! # Example
+//! ```toml
+//! [[block]]
+//! block  = "timewarrior"
+//! format = " $icon {$tags.str(w:8,rot_interval:4) $elapsed|}"
+//! ```
+//!
+//! # Icons Used
+//! - `tasks`
+
+use super::prelude::*;
+use chrono::DateTime;
+use tokio::process::Command;
+
+#[derive(Deserialize, Debug, SmartDefault)]
+#[serde(deny_unknown_fields, default)]
+pub struct Config {
+    #[default(30.into())]
+    interval: Seconds,
+    format: FormatConfig,
+
+    info: Option<u64>,
+    good: Option<u64>,
+    warning: Option<u64>,
+    critical: Option<u64>,
+}
+
+pub async fn run(config: &Config, api: &CommonApi) -> Result<()> {
+    let mut actions = api.get_actions()?;
+    api.set_default_actions(&[(MouseButton::Left, None, "stop_continue")])?;
+
+    let widget = Widget::new().with_format(config.format.with_default(" $icon {$elapsed|}")?);
+
+    loop {
+        let mut values = map! {
+            "icon" => Value::icon("tasks"),
+        };
+        let mut state = State::Idle;
+        let mut widget = widget.clone();
+
+        let data = get_current_timewarrior_task().await?;
+        if let Some(tw) = data {
+            if tw.end.is_none() {
+                // only show active tasks
+                let elapsed = chrono::Utc::now() - tw.start;
+
+                // calculate state
+                for (level, st) in [
+                    (&config.critical, State::Critical),
+                    (&config.warning, State::Warning),
+                    (&config.good, State::Good),
+                    (&config.info, State::Info),
+                ] {
+                    if let Some(value) = level {
+                        if (elapsed.num_minutes() as u64) >= *value {
+                            state = st;
```

Comment:
Instead of keeping a `state` variable and then later setting the widget state you can just set the widget's state directly.